Add reporting to goshawk

nikolas-burkoff opened the pull request:
Closes #145

npaszty commented:
@nikolas-burkoff 
great to see this functionality coming.

I submitted three issues to the reporter repo.

the existing issue to capture the encoding and filtering is to capture all of it. I could add manually to the card comment like "baseline constraint of <6.87 to < 20 but that's a major pain and I would like not do it correctly anyway because I'm just a human.

I realize that the baseline filter constraint is illustrated in the R code but what the user would want is to have a clear inventory of UI state that produced the output. is it possible to have an accompanying metadata card or something like that?

as far as layout just checking for understanding that each visualization tab would have it's own report? that would be less desirable than having one report for any modules plug and played into an app.

nikolas-burkoff commented:
> I realize that the baseline filter constraint is illustrated in the R code but what the user would want is to have a clear inventory of UI state that produced the output. is it possible to have an accompanying metadata card or something like that?

So in this PR we can easily add whatever accompanying information you would like for example [here](https://github.com/insightsengineering/teal.goshawk/blob/0de356d02d8511b848f89cb47dcb5e2120bb9b07/R/tm_g_gh_scatterplot.R#L386) I am manually adding the state of the filter panel. 

I can do things like this

```
card$append_text(paste("The opacity for the graph was chosen to be", input$opacity_value))
```

But I have to manually do this for every single UI component whose state you would like recorded (and if any UI changes I have to manually update it as well). 

There are a lot of UI components I cannot imagine you really need - either because they are already clearly on the visualisation (i.e. you know what the facet variable is, it's obvious) or who really cares (e.g. the slider value for the relative heights of the plot v the table- this does not materially affect the content of the output). So it would be good to know exactly which UI components are important enough to be added. 

If you really need _all_ of them (which isn't really the purpose of the report, that's the SRC) can you highlight the most important few to be done in this PR as crucial ones and we'll put another ticket in to add all the rest later - we have a _lot_ of modules to add reporting to this increment! 

> as far as layout just checking for understanding that each visualization tab would have it's own report? that would be less desirable than having one report for any modules plug and played into an app.

So it's one report for all visualisations - each time you click add to report you get a new card (i.e. section) of the same report and clicking download will give you a single pdf/html etc. containing all output from all cards

nikolas-burkoff commented:
So I've added the Data constraint to the report for the modules which use that as that seems like an obvious important one

![image](https://user-images.githubusercontent.com/15201933/177521701-d249a8ce-a01a-4719-bd5b-17c207491bfa.png)
![image](https://user-images.githubusercontent.com/15201933/177521738-f97d34f1-9ba4-4606-b68d-06b6cafa0fd2.png)
![image](https://user-images.githubusercontent.com/15201933/177521911-3cfb476d-6dc0-4a3e-ba71-84357596c154.png)

nikolas-burkoff commented on R/utils.R:
```diff
@@ -118,6 +118,21 @@ templ_ui_constraint <- function(ns, label = "Data Constraint") {
 
+# for outputting the constaint in the report
+formatted_data_constraint <- function(constraint_var, constraint_range_min, constraint_range_max) {
+  constraint_var_label <- switch(constraint_var,
+    "BASE2" = "Screening",
+    "BASE" = "Baseline",
+    "None"
```
So these are hard coded [here](https://github.com/insightsengineering/teal.goshawk/blob/e17fba3d652e6b24152efe9689bec279c2df9f31/R/utils.R#L84)

npaszty commented:
@nikolas-burkoff 
thanks for the explanations. thought there might be an easy way to just grab all in a list so as to minimize customization.
input$a = 
input$b = 
etc.

yes, the important/useful ones to capture are related to data filtering. so anything that isn't evident from the visualization itself. no need for purely aesthetic ui element state to be capture.
thanks for adding the constraint state.

did some more testing. this feature is fantastic! great enhancement. stakeholders will really love it I'm sure.

npaszty commented:
@nikolas-burkoff 

hope this approach to describe helps. I think the secondarys are helpful to stakeholders who may have a variety of skills to reproduce the visualizations. happy to chat further on this.

  - Boxplot: 
    - Primary: `Data Constraint` and `Facet by`
    - Secondary: `Select an X-Axis Variable` because instructional- and bound to the `Facet by` element to re-produce the visualization easily
  - Correlation:
    - Primary: `Data Constraint`
    - Secondary: `Treatment Variable Faceting` and `Regression Line` because they are instructional to re-produce the visualization easily
  - Density:
    - Primary: `Data Constraint`
    - Secondary: none
  - Lineplot:
    - Primary: `Data Constraint`, `Select Line Splitting Variable`, `Contributing Observations Threshold:`
    - Secondary: none
  - Spaghetti:
    - Primary:  `Data Constraint`
    - Secondary: none

nikolas-burkoff commented:
Thanks @npaszty I'll update this PR tomorrow.

At the moment I have "Data constraint" as a heading with the data constraint text below it.

- I could have a heading "Selected options" and within it add the data constraint and the other items?
- This could come before or after the visualisation in the report (or indeed the primary ones come before it and the secondary ones after if you want)
- Anything for the scatterplot, it's not on your list above?

npaszty commented:
@nikolas-burkoff 

I do like the idea of grouping the selecting input$* that we are capturing. agree on "Selected Options" and group under that. 
I think it's helpful to do all before each visualization card.

quite a while ago I mentioned to developer that correlation superseded scatter. thought it was going to be deprecated. if not then believe the same input$ we're capturing for correlation would be adequate.

personally I don't use scatter anymore since it is now a subset of correlation functionality. there are also other scatter modules in NEST.

Polkas commented on R/tm_g_gh_correlationplot.R:
```diff
@@ -868,6 +879,41 @@ srv_g_correlationplot <- function(id,
+    if (with_reporter) {
+      card_fun <- function(comment) {
+        card <- teal.reporter::TealReportCard$new()
+        card$set_name("Correlation plot")
+        card$append_text("Correlation plot", "header2")
```
```suggestion
        card$set_name("Correlation Plot")
        card$append_text("Correlation Plot", "header2")
```

please update each module to use Capital format for name and title

Polkas commented on R/tm_g_gh_correlationplot.R:
```diff
@@ -868,6 +879,41 @@ srv_g_correlationplot <- function(id,
+        card$append_text(
+          formatted_data_constraint(input$constraint_var, input$constraint_range_min, input$constraint_range_max)
+        )
+        card$append_text("Correlation plot", "header3")
```
```suggestion
        card$append_text("Plot", "header3")
```

We like your suggestion please use the Plot/Table for each subtitle for the main element.

mhallal1 commented on NEWS.md:
```diff
@@ -1,5 +1,8 @@
 # teal.goshawk 0.1.13.9002
 
+### Enhancements
+* Added `teal.reporter` reporting into the package modules/
```
```suggestion
* Added `teal.reporter` reporting into all package modules.
```

Polkas commented:
The teal.reporter is RELATIVELY much slower when we run goshawk app compared to exploratory, efficacy or safety.
We have to investigate it.

nikolas-burkoff commented:
@npaszty all additional options you requested have been added e.g.
![image](https://user-images.githubusercontent.com/15201933/177773510-7dad5054-258d-4d3e-9a12-04a27589d5fe.png)

@mhallal1 @Polkas I think I've addressed all your points as well - if everyone's happy I'll merge

nikolas-burkoff commented on R/tm_g_gh_correlationplot.R:
```diff
@@ -868,6 +879,48 @@ srv_g_correlationplot <- function(id,
+        card$append_text(
+          paste(
+            formatted_data_constraint(input$constraint_var, input$constraint_range_min, input$constraint_range_max),
+            "\nTreatment Variable Facetting:",
```
```suggestion
            "\nTreatment Variable Faceting:",
```

npaszty reviewed (approved):
@nikolas-burkoff 
reviewed and approve
👍
